docs/testplan/Generic-hash-test-plan.md

```diff
@@ -0,0 +1,314 @@
+5. Check the traffic is balanced over all the uplink ports.
+6. Randomly shutdown 1 nexthop interface.
+7. Send the traffic again.
+8. Check the traffic is balanced over all remaining uplink ports with no packet loss.
+9. Recover the interface and do shutdown/startup on the interface 3 more times.
+10. Send the traffic again.
+11. Check the traffic is balanced over all uplink ports with no packet loss.
+
+### Test cases #6 - Verify generic hash works properly when there are lag member flaps.
+1. The test is using the default links and routes in a t0/t1 testbed, and only runs on setups which have multi-member portchannel uplinks
+3. Configure all the supported hash fields for the ecmp and lag hash.
+4. Randomly select one hash field to test.
+5. If the hash field is DST_MAC, ETHERTYPE or VLAN_ID, change the topology to allow L2 switching and send L2 traffic in the next step.
+6. Send traffic with changing values of the field under test from a downlink ptf port to uplink destination.
+7. Check the traffic is balanced over all the uplink ports.
+8. Randomly shutdown 1 member port in all uplink portchannels.
+9. Send the traffic again.
+10. Check the traffic is balance over all remaining uplink ports with no packet loss.
+11. Recover the members and do shutdown/startup on them 3 more times.
+12. Send the traffic again.
+13. Check the traffic is balance over all uplink ports with no packet loss.
+
+### Test cases #7 - Verify generic hash running configuration persists after fast/warm reboot, and the saved configuration persists after cold reboot.
+1. The test is using the default links and routes in a t0/t1 testbed.
+2. Configure all the supported hash fields for the ecmp and lag hash.
+3. Randomly select one hash field to test.
+4. Randomly select a reboot type from fast/warm/cold reboot, if cold reboot, save the config before the reboot.
+5. Send traffic with changing values of the field under test from a downlink ptf port to uplink destination.
+6. Check the traffic is balance over all the uplink ports.
+7. Randomly do fast/warm/cold reboot.
+8. After the reboot, check the generic hash config via cli, it should keep the same as before the reboot.
+9. Send traffic again.
+10. Check the traffic is balance over all the uplink ports.
+
+### Test cases #8 - Verify the generic hash cannot be configured successfully with invalid parameters.
+1. Configure the ecmp/lag hash with invalid fields list parameter.
+2. Check there is a cli error that notifies the user the parameter is invalid.
+3. Check there is a warning printed in the syslog.
+4. Check the running config is not changed.
+5. The invalid parameters to test:
+  a. empty parameter
+  b. single invalid field
+  c. invalid fields combined with valid fields
+  d. duplicated valid fields(depends on the final implememtation)
+
+### Test cases #9 - Verify when a generic hash config key is removed, or updated with invalid values from config_DB via redis cli, there will be warnings printed in the log.
```

Comment:
@congh-nvidia do we really need to test this? The user interface is CLI only.

Reply:
This is required by the QG.

Reply:
@congh-nvidia got it.